File: merge-sort/merge-sort.js
// MergeSort
// Sorts the array by breaking it down into smaller chunks.

module.exports = function(array) {
  // Not an array, empty or array of 1 is already sorted
	if(!Array.isArray(array) || array.length < 2)
		return array;

	function merge(left, right) {
		var result = [];
		while(left.length || right.length) {
			if(left.length && right.length) {
				if(left[0] < right[0]) {
					result.push(left.shift());
				} else {
					result.push(right.shift());
				}
			} else if (left.length) {
				result.push(left.shift());
			} else {
				result.push(right.shift());
			}
		}
		return result;
	}

	function MergeSort(array){
		var length = array.length,
		mid = Math.floor(length * 0.5),
		left = array.slice(0, mid),
		right = array.slice(mid, length);
		if(length === 1) {
			return array;
		}
		return merge(MergeSort(left), MergeSort(right));
	}

	return MergeSort(array);
}

File: merge-sort/merge-sort.rb
def merge_sort(a)
  return a if a.size <= 1
  l, r = split_array(a)
  combine(merge_sort(l), merge_sort(r))
end

def split_array(a)
  mid = (a.size / 2).round
  [a.take(mid), a.drop(mid)]
end

def combine(a, b)
  return b.empty? ? a : b if a.empty? || b.empty?
  smallest = a.first <= b.first ? a.shift : b.shift
  combine(a, b).unshift(smallest)
end

# a = [6, 23, 53, 1, 2, 5, 62, 61, 33, 21, 14, 6, 23].shuffle
# p merge_sort(a)
